packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1355,29 +1287,9 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
     }
 
     private async summarizeInternal(fullTree: boolean, trackState: boolean): Promise<ISummarizeInternalResult> {
-        // A list of this channel's GC nodes. Starts with this channel's GC node and adds the GC nodes all its child
-        // channel contexts.
-        let gcNodes: IGraphNode[] = [ this.getGCNode() ];
         const builder = new SummaryTreeBuilder();
 
-        // Iterate over each store and ask it to snapshot
-        await Promise.all(Array.from(this.dataStores.contexts)
-            .filter(([_, context]) => {
-                // Summarizer works only with clients with no local changes!
-                assert(context.attachState !== AttachState.Attaching);
-                return context.attachState === AttachState.Attached;
-            }).map(async ([contextId, context]) => {
-                const contextSummary = await context.summarize(fullTree, trackState);
-                builder.addWithStats(contextId, contextSummary);
-
-                // back-compat 0.30 - Older versions will not return GC nodes. Set it to empty array.
-                if (contextSummary.gcNodes === undefined) {
-                    contextSummary.gcNodes = [];
-                }
-
-                // Update and add the child context's GC nodes to the main list.
-                gcNodes = gcNodes.concat(this.updateChildGCNodes(contextSummary.gcNodes, contextId));
-            }));
+        const gcNodes = await this.dataStores.summarizeInternal(builder, fullTree, trackState);
```

Comment:
This does not look right. dataStores.summarizeInternal should return ISummarizeInternalResult. The container blobs can then be added to it directly without needing a builder.

Reply:
The SummaryTreeBuilder can then move in to dataStores.

Reply:
i don't really like the api, but there are other things done in the container runtime with the SummaryTreeBuilder see blobs below, and i'm trying not to introduce new logic here. specifically i don't want to merge ISummarizeInternalResults.

Reply:
I think something like that is the right next step. Ideally, we have a more structured pattern for hooking thing up to the summary which is currently deeply integrated into the data store contexts, but then overlays dds.

Reply:
I don't mind the SummaryTreeBuilder so much. But summarizeInternal returning gcBlobs makes it inconsistent with rest of summary APIs.
Can you change dataStore's summarizeInternal to not return anything for now? And return gcBlobs as an empty array from here.
I will update of all this in my upcoming change that actually uses the gcNodes.

Reply:
i don't see another obvious way to extract this without change, but open to ideas if i am missing something.

Reply:
I think we replied at the same time :-). Take a look at my suggestion if it makes sense. I will update this to a more structured pattern like you mentioned later.

Reply:
I think this is probably fine, but also agree the API of relying on mutating your arg is uncomfortable.
I think I can get more behind it when I think about the ContainerRuntime is summarizing a bunch of different data, not just the data stores, so each part it is summarizing just attaches its subtrees and blobs to the builder tree.  The only thing is I might like to see a different name for `DataStores.summarizeInternal` so it is a little more clear? But not sure what that should be. Like add data store subtrees or something.

Reply:
the longer-term goal is to align this with a similar api for dds, or as @vladsud would say "Everything is a channel". This opens up a bunch of scenarios. Short term, i just wanted to clean up the container runtime a bit, and hopefully make it a bit more testible. I think the summarization api is a good candidate to look at next, as ideally, we can make it a bit more testible as an isolated unit, as if it's just different parts attaching sub-trees, those parts could in theory just be test things, rather than real things like data stores or blobs.